### PR TITLE
New wizard spell - Ring of Fire

### PR DESCRIPTION
**Pull request by unid15**
Adds a new spell - Ring of Fire

Cooldown: 30 seconds (becomes 10 seconds if upgraded 3 times)
Duration: 10 seconds
When cast, summons a **stationary** ring of fire with a radius of 3 turfs. All mobs that touch it receive damage and are set on fire (receiving more damage until they are extinguished). A mob that stays in the ring for the whole duration receives around 80-90 burn damage
While the ring is active, the wizard is immune to fire and heat.
You can upgrade the spell to make the ring always centered on you.
The ring is powered by magic so it's immune to vacuum, cold and fire extinguishers

Possible synergies:
Magic Missile + Ring of Fire = stun everybody and burn them all at once
Ethereal Jaunt + Ring of Fire 2 = become a flying ring of death

rof

Also fixes heat immune mobs getting damaged by flamethrowers / fire breath attacks

TODO:
- [x] Check if locking categories are working correctly
- [x] Test for bugs

### CODE REVIEW

**Comment by clusterfack** [comment]
WEAPONIZED PUNS

**Comment by PJB3005** [comment]
For the unique thing with the locking categories, you should definitely use strings as ID for the locking category to make them unique.

**Comment by Kurfursten** [comment]
I think this may be a bug. When you moved down, the flames also moved down, but I believe that as you go down the flames should go ever higher.

**Comment by unid15** [comment]
I don't get it 😐

**Comment by PJB3005** [comment]
@unid15 I made sure to think about the need for dynamically generated categories, so `lock_atom()` can take more than typepaths, it can also take strings.

If you give it a string as category you can have unique categories without the same ID, because the way you've got it set up right now makes everything a mess, you're making duplicate categories with the same IDs and the system can't distinguish between them.

I did forget to make a proc to add categories of a certain type with a certain ID, however. So you'll have to do that, it'd look something like this:

``` DM
// Proc for adding an unique locking category with a certain ID.
/atom/movable/proc/add_lock_cat(var/type, var/id)
    if (id in locking_categories_name)
        return locking_categories_name[id]

    var/datum/locking_category/C = getFromPool(type, src)
    C.name = id
    locking_categories_name[id] = C
    locking_categories += C
```

Basically, for each thing you want to lock:
- Register a new category with the above proc (pass a unique ID for each category as second argument!).
- Use the unique category to lock the individual fires.

**Comment by Kurfursten** [comment]
@unid15 https://www.youtube.com/watch?v=It7107ELQvY

File: code/modules/spells/general/ring_of_fire.dm [hunk 1]
```diff
@@ -5,10 +5,10 @@
 	charge_max = 300
 	cooldown_min = 100
 
-	spell_levels = list(Sp_SPEED = 0)
-	level_max = list(Sp_TOTAL = 3, Sp_SPEED = 3)
+	spell_levels = list(Sp_SPEED = 0, Sp_POWER = 0)
+	level_max = list(Sp_TOTAL = 3, Sp_SPEED = 3, Sp_POWER = 0)
 
-	spell_flags = NEEDSCLOTHES
+	spell_flags = NEEDSCLOTHES | Z2NOCAST
```
Comment (ComicIronic) [inline code/modules/spells/general/ring_of_fire.dm]: But why

**Comment by Blithering** [comment]
I would prefer if the fire conjured by the spell stayed in place originally, with an upgrade that makes it follow the caster as it does now. being able to set the entire crew on fire through walls with jaunt seems a bit much for a two spell combo.

**Comment by Aurx** [comment]
Agreed, jauntfire would become a new meta overnight.

**Comment by icantthinkofanameritenow** [comment]
I'm sure we could do better than literally copying one of Polaris's spells.  Seriously, this is exactly the same as one of their "Technomancer" spells. Plagiarism doesn't help anyone, especially when it stinks of not-Bay.

**Comment by ComicIronic** [comment]
> Plagiarism doesn't help anyone

Unless it's plagiarism of good content.

**Comment by ghost** [comment]
I mean, we've plagiarised TG changeling content before and nobody has complained too much, icant :^)

**Comment by unid15** [comment]
I didn't get the idea from Polaris though, Ring of Fire seems like a generic enough D&D spell




I'll try to balance it further once the spellbook refactor is in (as it allows having multiple upgrade trees instead of just cooldown and power)

**Comment by Exxion** [comment]
Our licenses are compatible
We couldn't actually plagiarize from them if we wanted

**Comment by Intigracy** [comment]
You should make it an actual ring instead of a box.

circles3

**Comment by ihadtoregisterforthis** [comment]
But it is
image

**Comment by Shadowmech88** [comment]
A one-tile-thick ring seems like it'd be easy for people to avoid direct contact with it.

Maybe in addition to speed upgrades, the wizard could choose power upgrades that would make the ring thicker?

**Comment by unid15** [comment]
Balance changes:
- The ring is now stationary unless upgraded
- While the ring is active, the caster is immune to fire and heat (to prevent them from burning on their own ring)
- Increased tick damage by 1

File: code/modules/spells/general/ring_of_fire.dm [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+			to_chat(L, "<span class='info'>You feel resistant to fire.</span>")
+
+		L.mutations.Add(M_RESIST_HEAT)
+		L.update_mutations()
+
+		spawn(duration)
+			L.mutations.Remove(M_RESIST_HEAT)
+			L.update_mutations()
+
+			if(!L.mutations.Find(M_RESIST_HEAT))
```
Comment (ghost) [inline code/modules/spells/general/ring_of_fire.dm:46]: Uh, why would you still have it?

File: code/modules/spells/general/ring_of_fire.dm [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+	to_chat(user, "<span class='danger'>You summon a ring of fire around yourself.</span>")
+
+	if(isliving(user))
+		var/mob/living/L = user
+
+		if(!L.mutations.Find(M_RESIST_HEAT))
+			to_chat(L, "<span class='info'>You feel resistant to fire.</span>")
+
+		L.mutations.Add(M_RESIST_HEAT)
```
Comment (Exxion) [inline code/modules/spells/general/ring_of_fire.dm:39]: Shouldn't this and the following code be tabbed so you don't lose heat resistance if you already had it?

File: code/modules/spells/general/ring_of_fire.dm (hunk 2 above)
Comment (unid15) [reply code/modules/spells/general/ring_of_fire.dm:46]: The heat resistance gene, for example. Other sources of heat immunity too, if there are any.

File: code/modules/spells/general/ring_of_fire.dm (hunk 3 above)
Comment (unid15) [reply code/modules/spells/general/ring_of_fire.dm:39]: If you already have heat resistance, your mutations list will have a duplicate M_RESIST_HEAT in it. That duplicate is then removed after 10 seconds.

If that was tabbed, you'd lose heat resistance from eating ryetalin (if you had the heat resistant gene beforehand, which gives you the same mutation)

Comment (Exxion) [reply code/modules/spells/general/ring_of_fire.dm:39]: Oh, you can have duplicate mutations?
Wow

**Comment by Arthurdentist** [comment]
If possible could you add single use spellbooks for this spell and your other spells?

Spellbook bundle automatically adds all spellbooks to its list so its just a matter of making the single use book and coming up with some nasty surprise for anyone who uses it after its been used.

**Comment by Shadowmech88** [comment]
The ring of fire book could just set the person on fire if they try to use it a second time.

**Comment by clusterfack** [comment]
I don't like the fire sprite for this personally but who cares.

Anyway, what could go wrong?
